### PR TITLE
Support parsing modules

### PR DESCRIPTION
**Pull request by moticless**
- Support parsing modules.
- Fix bug in rdbLoadLen() when it reads 64bits
- Support parsing module-auxiliary data 
- Adding deps/redis/config.h for BYTE_ORDER correctness
- Add hiredis as submodule for testing
- Utilize hiredis for the tests
- RDB2RESP: Extend API with flag supportRestoreModuleAux

### CODE REVIEW

File: src/lib/parser.c [hunk 1]
```diff
@@ -1678,8 +1682,103 @@ RdbStatus elementFunction(RdbParser *p) {
...
+            }
+            default:
+                RDB_reportError(p, RDB_ERR_MODULE_INVALID_STATE,
+                    "elementModule() : Invalid parsing element state: %d.", ctx->state);
+                return RDB_STATUS_ERROR;
+        }
+    }
+
+    /* Static assert that verifies enums are aligned */
+    _Static_assert(ST_MODULE_OPCODE_SINT==RDB_MODULE_OPCODE_SINT, "enums must be aligned");
```
Comment (tezc) [inline src/lib/parser.c:1759]: `_Static_assert` was introduced in c11 if I'm not mistaken. gcc is generally doing terrible about warning these issues but clang would complain most probably.

Comment (moticless) [reply src/lib/parser.c:1759]: Let's refactor such kind of code when we goanna have CI that support clang to verify it.

Comment (tezc) [reply src/lib/parser.c:1759]: Just fyi, In my experience, compilers don't warn you reliably about the standard (it should be an easy thing to do though). Generally, you don't know until compile fails on some platform. So, I try to avoid features if I know it does not exist in a standard or on a platform. 

Anyway, here it is clang build, maybe it will throw some warnings: https://github.com/redis/librdb/pull/18

File: src/lib/parserRaw.c [hunk 2]
```diff
@@ -583,6 +583,119 @@ RdbStatus elementRawSet(RdbParser *p) {
...
+        ST_RAW_MODULE_GET_ID=0,
+
+        /* Following enums are aligned to module-opcodes. Static assert below. */
+        ST_RAW_MODULE_OPCODE_SINT,
+        ST_RAW_MODULE_OPCODE_UINT,
+        ST_RAW_MODULE_OPCODE_FLOAT,
+        ST_RAW_MODULE_OPCODE_DOUBLE,
+        ST_RAW_MODULE_OPCODE_STRING_CALL_STR,
+
+        ST_RAW_MODULE_OPCODE_STRING_STR_RETURN, /* return from sub-element string parsing */
```
Comment (oranagra) [inline src/lib/parserRaw.c:597]: i don't understand (yet) this mechanism of CALL_STR and STR_RETURN.
maybe it's explained in code i didn't yet get to.

File: src/lib/parserRaw.c [hunk 3]
```diff
@@ -583,6 +583,119 @@ RdbStatus elementRawSet(RdbParser *p) {
...
+                return subElementCall(p, PE_RAW_STRING, ST_RAW_MODULE_OPCODE_STRING_STR_RETURN);
+            }
+
+            case ST_RAW_MODULE_OPCODE_STRING_STR_RETURN: {
+                /*** ENTER SAFE STATE (no rdb read)***/
+                size_t len;
+                char *dataRet;
+                subElementCallEnd(p, &dataRet, &len);
+                updateElementState(p, ST_RAW_MODULE_NEXT_OPCODE);
+                break;
```
Comment (oranagra) [inline src/lib/parserRaw.c:662]: what's that safe state? safe for what and when does it end?

File: test/dumps/module_data.json [hunk 4]
```diff
@@ -0,0 +1 @@
+"key1":"<<BINARY>>"
```
Comment (oranagra) [inline test/dumps/module_data.json:1]: maybe we can print the module name and maybe size of the payload?

File: test/test_common.c [hunk 5]
```diff
@@ -216,19 +206,38 @@ void setupRedisServer() {
...
+        /* if module testrdb.so exists (ci.yaml takes care to build testrdb), part
+         * of redis repo testing, then load it for test_rdb_to_redis_module. The
+         * test will run only if testrdb appear in the server "MODULE LIST",
+         * otherwise skipped gracefully. */
+        if (access(testrdbModulePath, F_OK) != -1) {
+            execl(fullpath, fullpath,
+                  "--port", redisPortStr,
+                  "--dir", "./test/tmp/",
+                  "--logfile", "./redis.log",
+                  "--loadmodule", testrdbModulePath,
```
Comment (oranagra) [inline test/test_common.c:226]: do we have a test that uses RESTORE command and validates the module data was loaded on the dest correctly?

File: src/lib/parserRaw.c (hunk 2 above)
Comment (moticless) [reply src/lib/parserRaw.c:597]: Available for a while and exaplained in the code next to function `subElementCall` and struct `ParsingSubElement` :
```
/****************************************************************
 * Sub-Element parsing
 *
 * The parser can handle one level of nested parsing-elements (PE), whereby a PE
 * may be called by another PE and control is returned to the caller once the
 * parsing of sub-element is complete.
 *
 * Currently, this functionality is only adapted and utilized by raw string element
 * which can also run as sub-element of other data-types. It takes care to write the
 * processed string to raw-aggregator, as part of RESTORE of current command.
 *
 * In addition it returns the string decoded by allocating bulkUnmanagedAlloc()
 * (The caller won't need to release returned data yet restrictd to use it within
 * its current parsing-element state).
 ****************************************************************/
```

File: src/lib/parserRaw.c (hunk 3 above)
Comment (moticless) [reply src/lib/parserRaw.c:662]: Available for a while and exaplained in the README. Please see section `Parsing-Element states` and subsection ` Defining Safe-state`. Please let me know if it is clear enough.

File: test/dumps/module_data.json (hunk 4 above)
Comment (moticless) [reply test/dumps/module_data.json:1]: Maybe we can optionally return some kind of iterator (for int, float, double, double & string) and let handlers to decide what they want to do with it ... But then the handlers might need to deal with `RDB_STATUS_WAIT_MORE_DATA`. If so, then it will be simpler to have another group of callbacks to submodules (that called only after each item was entirely read) and let the handlers do the math, if it wants the size.

File: test/test_common.c (hunk 5 above)
Comment (moticless) [reply test/test_common.c:226]: See function `test_rdb_to_redis_module`. It might be little fragile though. As precondition, the test verifies that testrdb module is loaded into redis, otherwise the test gracefully skipped.

(Note that all quick and dirty `redis-cli` system calls should be replaced with integrtaing `hiredis` library eventually)

File: src/lib/parserRaw.c (hunk 2 above)
Comment (oranagra) [reply src/lib/parserRaw.c:597]: so for module data types, you this sub-element for strings, and but not for other primitives?
is that described somewhere?
i understand it let's you break big module payloads to many individual calls (even split the string to several calls and avoid blocking IO), right?
what would happen if a module payload is composed of a ton of integers (let's say 100mb worth of integers)?

File: test/dumps/module_data.json (hunk 4 above)
Comment (oranagra) [reply test/dumps/module_data.json:1]: IIUC, you're suggesting that users can implement extensions to the parser that can decode module keys and translate them to usable content (e.g. for ReJson, they'll be able to print the JSON content of the key).
that sounds like a good idea, maybe we can add that in a todo and handle it in a second phase.

meanwhile, for someone that dumps an RDB into json and some keys are module keys, i think just saying "BINARY" is insufficient, and i'd like to state the module name and maybe the payload size.

File: test/test_common.c (hunk 5 above)
Comment (oranagra) [reply test/test_common.c:226]: great.. why skip it if the module load fails? i'm afraid we'll miss it instead of fixing the problem / improving the test.

File: src/lib/parser.c (hunk 1 above)
Comment (moticless) [reply src/lib/parser.c:1786]: As there is a lot of vudu code to have cross-version code for assert, maybe we can limit static-assert just for some of the versoins which is good enough.

File: src/lib/parserRaw.c (hunk 2 above)
Comment (moticless) [reply src/lib/parserRaw.c:587]: Right.

It will call inefficiently to string-element multiple times through `parserMainLoop()`. Those inefficient calls can be optimized by making direct call and avoid from passing through `parserMainLoop()`. For example, instead of : 
```
            case ST_RAW_MODULE_OPCODE_STRING_CALL_STR: {
                /* call raw string as subelement */
                return subElementCall(p, PE_RAW_STRING, ST_RAW_MODULE_OPCODE_STRING_STR_RETURN);
```

We shall have:
```
            case ST_RAW_MODULE_OPCODE_STRING_CALL_STR: {
                /* call raw string as subelement */
                subElementCall(p, PE_RAW_STRING, ST_RAW_MODULE_OPCODE_STRING_STR_RETURN);
                IF_NOT_OK_RETURN(elementRawString(p)); 
                break;   /* short transition between module states without passing through main-loop */

```

I should reach later to performance phase. Yet to plan what is goanna be tested.

Comment (moticless) [reply src/lib/parserRaw.c:587]: Maybe I missed your point. If it is native module integer then it won't call to string and the transition between the states of the module will be made only within element-module boundary (i.e. function `elementRawModule()`)

File: test/test_common.c (hunk 5 above)
Comment (moticless) [reply test/test_common.c]: ci.yml should fail if failed to build `testrdb.so`. If we reach here and the file exists then for sure the test will capture the fail.
The graceful fallback to the case `testrdb.so` doesn't exists is to be able, later on, to run these tests against earliest redis version that I can reach. In addition there is the option to expose the tests for external redis server which can assist debugging Redis side in the future.

File: src/lib/parserRaw.c (hunk 2 above)
Comment (oranagra) [reply src/lib/parserRaw.c:587]: my last question was if it'll break out in between module integers and return to the caller?
and for string we use the sub-element in order to break if further (i.e. break the string into smaller chunks)?

File: test/test_common.c (hunk 5 above)
Comment (oranagra) [reply test/test_common.c]: it's a little odd to depend on the module from redis, and even odder that it's loosely based on it (graceful failures).
maybe we should bring that module (with some simplifications) into this repo and avoid that dependency.

File: src/lib/parser.c (hunk 1 above)
Comment (tezc) [reply src/lib/parser.c:1786]: not sure I understand. Do you want to keep static assert? If so, this is how it is defined on my platform in cdefs.h: 

```c
#define _Static_assert(expr, diagnostic) \
    extern int (*__Static_assert_function (void)) \
      [!!sizeof (struct { int __error_if_negative: (expr) ? 2 : -1; })]
```

I guess you can define this if undefined. 


Btw, maybe you can assign those defines directly and avoid checking it altogether:

```c
enum {
    ST_MODULE_OPCODE_SINT = RDB_MODULE_OPCODE_SINT 
    ...
}
```

Comment (oranagra) [reply src/lib/parser.c:1786]: i wanna remind us that these values are part of the rdb format and are never gonna change.
it's not like some other (volatile-only) enum, that can change but two components must remain allied.
it's like asserting that TCP is protocol number 6 in the IP packet header.
not needed IMHO.

Comment (moticless) [reply src/lib/parser.c:1786]: @tezc, 
I wrote it first like that but then I thought about the issue that enum doesn't verifies duplication of values.
But then thinking about it again, it might be good compromise between readability of the code and alignment enforcement. I think it I will go with your suggestion. 

@oranagra,
The verification is around changing the states values of the element that might break the alignment with the rdb opcodes. Not to verify that rdb opcodes has suddenly changed.

File: src/lib/parserRaw.c (hunk 2 above)
Comment (moticless) [reply src/lib/parserRaw.c:587]: Each new (sub)element+state is being updated and can reinvoked in case of early return. 
Most of the librdb tests run also in a mode that fails all "new" read and simulates RDB_STATUS_WAIT_MORE_DATA from RDB reader/buffer, which enforces the parser to return from its current "state" and reinvoke again.

File: src/ext/handlersToResp.c [hunk 6]
```diff
@@ -380,73 +382,131 @@ static RdbRes toRespFunction(RdbParser *p, void *userData, RdbBulk func) {
...
+    /* reset rawCtx */
+    RdbxToResp *ctx = userData;
+    ctx->rawCtx.rawSize = size;
+    ctx->rawCtx.sentFirstFrag = 0;
+    ctx->rawCtx.isModuleAux = 1;
+    ctx->rawCtx.crc = 0;
+
+    /* if target doesn't support module-aux, then skip it */
+    if (!ctx->conf.supportRestoreModuleAux)
+        return RDB_OK;
```
[the remark this answers is not in the document]
Comment (moticless) [reply src/ext/handlersToResp.c:398]: It will add more complexity to responses pipeline tracking that I am not sure if it is worth it.
Another option is to extend boolean variable `supportRestore` to be:
 0=No support, 1=Support restore, 2=Support Restore with Module Aux

In the tradeoff of spamming the API with weird flags and deep hiding such feature, this sound 
most reasonable to me.

File: src/ext/handlersToResp.c [hunk 7]
```diff
@@ -380,73 +382,131 @@ static RdbRes toRespFunction(RdbParser *p, void *userData, RdbBulk func) {
 
 /*** Handling raw ***/
 
+static RdbRes toRespRawBeginModuleAux(RdbParser *p, void *userData, RdbBulk name, int encver, int when, size_t size) {
```
[the remark this answers is not in the document]
Comment (moticless) [reply src/ext/handlersToResp.c:385]: Few lines below it can be seen assigned to `ctx->rawCtx.rawSize = size;`
I will modify it to `rawSize`

File: src/ext/handlersToResp.c (hunk 6 above)
[the remark this answers is not in the document]
Comment (oranagra) [reply src/ext/handlersToResp.c:398]: i prefer not to add an interface at all (just silently ignore), but if we don't want that, then a separate interface would be easier to remove later (rather than mixing it with another interface).

i think at some point you'll have to have the complexity of understanding responses and associating them with the commands you sent anyway, and once you do that, then this optimistic approach won't add a big overhead.

one example is that when a command fails, i think we must log which command it was, not just the error message, otherwise, trouble shooting will be very difficult.

File: src/ext/handlersToResp.c (hunk 7 above)
[the remark this answers is not in the document]
Comment (oranagra) [reply src/ext/handlersToResp.c:385]: but how do we know the rawSize when we begin a module aux? where do you take that info from?
i have a feeling (didn't bother to look), that it's always 0, and not used in this case...

File: src/ext/handlersToResp.c (hunk 6 above)
[the remark this answers is not in the document]
Comment (moticless) [reply src/ext/handlersToResp.c:398]: Note that librdb can generate RESP output. How we suppose to decide whether to generate this special command or not? in that case, I prefer to leave the flag as is.

BTW, I do have some basic mechanism for pipeline to report of which command caused the error. Example to 100 pending list commands and failure on `mylist62` because the test makes "SET mylist62 1" before
```
:: ERROR :: [elementQuickList::State=1] 
Received Server error: "WRONGTYPE Operation against a key holding the wrong kind of value"
Got failed on command [#9501] (First 40 bytes):
*3
$5
RPUSH
$8
mylist62
$19
0.035

Preceding command [#9500] was: 
*3
$5
RPUSH
$8
mylist30
$17
0.791
```

File: src/ext/handlersToResp.c (hunk 7 above)
[the remark this answers is not in the document]
Comment (moticless) [reply src/ext/handlersToResp.c:385]: Please read the comment:
```
typedef struct RawContext {

    /* aggType - The parser encounters difficulty determining the size of
     * certain types during their parsing because they lack a preceding header
     * that would indicate their size. This creates a problem when implementing
     * callback handlers for executing RESP2 RESTORE commands on a live server,
     * which requires the type's size to be sent at the beginning. To resolve
     * this issue, the parser gathers all payload data for these types and only
     * provides it to callback handlers once it reaches the end of the type and
     * knows its total size. However, for types like strings, whose sizes are
     * already known at the beginning, the parser will not aggregate the entire
     * payload if it is large enough, but stream it to handleFrag callback. */
    enum {
        AGG_TYPE_UNINIT,
        AGG_TYPE_ENTIRE_DATA,
        AGG_TYPE_PARTIALLY
    } aggType;
```

File: src/ext/handlersToResp.c (hunk 6 above)
[the remark this answers is not in the document]
Comment (oranagra) [reply src/ext/handlersToResp.c:398]: good point about sending RESP to a file rather than redis.
but then i wonder what do people do with that file? if they pipeline it to redis later (e.g. with `netcat`), then maybe they ignore replies (and errors) anyway.

regarding sending it to a live server, good that you have that feature in the error report.. but that means that it's not too complicated to have it implicitly ignore some errors (like unknown command, only on a certain command).

anyway, my point is that at some point in time we'll want to drop that interface, so i rather not have it if we can avoid it.

File: src/ext/handlersToResp.c (hunk 7 above)
[the remark this answers is not in the document]
Comment (oranagra) [reply src/ext/handlersToResp.c:385]: ok, i didn't remember the difference between this parser and the one i was involved with.
i.e. in the other parser the aggregation was on the receiver's side, so when starting a new payload the size wasn't known.

So here, it aggregates the entire thing before calling "Begin", and then it repeatedly calls the "Frag" callback? (i suppose that's necessary since each of them may be sitting in a separate allocation).

so this one doesn't have the optional size hint callback, and the "First" frag concept is internal to the callee, right?

thanks for clearing this up...

[the remark this answers is not in the document]
Comment (moticless) [reply src/ext/handlersToResp.c:385]: Exactly

File: test/test_common.c (hunk 5 above)
Comment (moticless) [reply test/test_common.c]: Note that compiling module requires to have redis source. I prefer to leave it as is quick and dirty (?!). And when we reach to a point that we want to expand the support in modules, then make the effort. 
Note that graceful failure is when the `testrdb.so` is not available. In our CI I will take care that it will verify success of `make -C tests/modules`

File: .gitmodules [hunk 8]
```diff
@@ -0,0 +1,3 @@
+[submodule "deps/hiredis"]
+	path = deps/hiredis
+	url = https://github.com/redis/hiredis.git
```
Comment (tezc) [inline .gitmodules:3]: Maybe not part of this PR but do you think we can use subtree instead of submodule? With subtree, dependency project code is always in the repo, no need to run extra commands to fetch/clear it. (As I mentioned before, I also feel the same way for cmocka as well, it should be part of the repo to make the developers' life easier)

Comment (moticless) [reply .gitmodules:3]: I am not much familiar with subtree. Sound like a good idea. Let's do it on a distinct PR.

File: test/test_common.c (hunk 5 above)
Comment (oranagra) [reply test/test_common.c]: compiling a module does NOT require the redis source, only redismodule.h, which is commonly just copied into the module's source tree.
i think your test better be self-sufficient and have their own module (copied from redis and cleaned).
you can leave that for a later step, but i think that's the right approach.

Comment (moticless) [reply test/test_common.c]: I agree with you that this is the right approach... I am too lazy about it because I know we goanna  expand it and write it properly (It can be valuable feature for modules)

**Review by YaacovHazan — APPROVED** [review]

**Review by YaacovHazan — APPROVED** [review]

**Review by YaacovHazan — APPROVED** [review]

**Review by YaacovHazan — APPROVED** [review]

**Review by YaacovHazan — APPROVED** [review]

**Review by YaacovHazan — APPROVED** [review]

**Review by YaacovHazan — APPROVED** [review]

File: .github/workflows/ci.yml [hunk 9]
```diff
@@ -25,17 +27,18 @@ jobs:
           ctest
           sudo make install
           # Install redis for tests against external Redis server
-          cd ~
-          git clone https://github.com/redis/redis.git
-          cd redis 
-          make -j 4
-
+                              
+          # Build Redis to run it against librdb tests
```
Comment (yossigo) [inline .github/workflows/ci.yml:31]: @moticless You may want to consider getting a Redis binary, like we do [here](https://github.com/RedisLabs/memtier_benchmark/blob/4203084b085e0f93bb9130c281362ea651d2140c/.github/workflows/ci.yml#L29). It's potentially faster.

**Review by yossigo — APPROVED** [review]

File: .github/workflows/ci.yml (hunk 9 above)
Comment (moticless) [reply .github/workflows/ci.yml:31]: Good reference. thanks.

Currently librdb is relying on Redis repo for building redis binary and reusing `testrdb.c` source to test modules parsing.
I know it is preferable to librdb have its own source to build testable module but i prefer to wait with that effort since I plan anyway (second phase?!) to expand callback module API to parse primitives as well (and not only RESTORE module).
All in all, additional  +1minute of CI time.
